### PR TITLE
[MIG] product_multi_barcode: Migration to 16.0

### PR DESCRIPTION
**Pull request by StefanRijnhart**
From unmerged https://github.com/OCA/stock-logistics-barcode/pull/434

This PR uncovers a problem with `odoo-test-helper` used in `barcodes_generator_abstract`'s tests. This should be fixed in https://github.com/OCA/odoo-test-helper/pull/22.

### CODE REVIEW

**Comment by rousseldenis** [comment]
@StefanRijnhart Still draft ?

**Comment by StefanRijnhart** [comment]
@rousseldenis ~~waiting for the odoo-test-helper fix to get this one green. Thanks for checking that one!~~ Now ready for review.

**Comment by StefanRijnhart** [comment]
Pre-commit fixed here: https://github.com/OCA/stock-logistics-barcode/pull/491

File: product_multi_barcode/models/product_barcode.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            rec.product_id = rec.product_tmpl_id.product_variant_ids[0]
+
+    @api.constrains("name")
+    def _check_duplicates(self):
+        for record in self:
+            barcodes = self.search(
+                [("id", "!=", record.id), ("name", "=", record.name)]
+            )
+            if barcodes:
+                raise UserError(
```
Comment (rousseldenis) [inline product_multi_barcode/models/product_barcode.py]: You should use ValidationError in constraints

File: product_multi_barcode/models/product_product.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    @api.depends("barcode_ids.name", "barcode_ids.sequence")
+    def _compute_barcode(self):
+        for product in self:
+            product.barcode = product.barcode_ids[:1].name
+
+    def _inverse_barcode(self):
+        for product in self:
+            if product.barcode_ids:
+                product.barcode_ids[:1].write({"name": product.barcode})
```
Comment (rousseldenis) [inline product_multi_barcode/models/product_product.py]: I think write() is not correct in an inverse method. You should assign value instead.

File: product_multi_barcode/models/product_product.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            elif not product.barcode:
+                product.barcode_ids.unlink()
+            else:
+                self.env["product.barcode"].create(self._prepare_barcode_vals())
```
Comment (rousseldenis) [inline product_multi_barcode/models/product_product.py]: Gather barcode vals in a list, then use multi create.

File: product_multi_barcode/models/product_product.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    def _compute_barcode(self):
+        for product in self:
+            product.barcode = product.barcode_ids[:1].name
+
+    def _inverse_barcode(self):
+        for product in self:
+            if product.barcode_ids:
+                product.barcode_ids[:1].write({"name": product.barcode})
+            elif not product.barcode:
+                product.barcode_ids.unlink()
```
Comment (rousseldenis) [inline product_multi_barcode/models/product_product.py]: Gather barcode ids in a set(), then barcode_model.browse(ids).unlink()

File: product_multi_barcode/tests/test_product_multi_barcode.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo.tests import TransactionCase, tagged
+
+from ..hooks import post_init_hook
+
+
+@tagged("post_install", "-at_install")
+class TestProductMultiBarcode(TransactionCase):
+    def setUp(self):
```
Comment (rousseldenis) [inline product_multi_barcode/tests/test_product_multi_barcode.py]: Could you change this to classmethod ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review. Some performance improvements can be done.

File: product_multi_barcode/models/product_barcode.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    def _compute_product(self):
+        for rec in self.filtered(
+            lambda x: not x.product_id and x.product_tmpl_id.product_variant_ids
+        ):
+            rec.product_id = rec.product_tmpl_id.product_variant_ids[0]
+
+    @api.constrains("name")
+    def _check_duplicates(self):
+        for record in self:
+            barcodes = self.search(
```
Comment (hieulucky111) [inline product_multi_barcode/models/product_barcode.py]: should use `search_count`?

File: product_multi_barcode/models/product_product.py (hunk 3 above)
Comment (hieulucky111) [inline product_multi_barcode/models/product_product.py]: should be `product._prepare_barcode_vals()`

**Review by hieulucky111 — CHANGES_REQUESTED** [review]

**Review by aliciagaarzo — CHANGES_REQUESTED** [review]
Review comments and cover all cases in tests.

**Comment by hugo-cordoba** [comment]
Hello @rousseldenis , I have seen that @aliciagaarzo  has made the changes you requested, how do you see them? I think the development is correct, can we help to merge the code? Thank you very much, have a nice day.

**Comment by liebana** [comment]
I've made a functional review and LGTM.

Friendly ping to @rousseldenis @StefanRijnhart @hieulucky111

**Review by liebana — APPROVED** [review]

**Comment by rousseldenis** [comment]
@StefanRijnhart Could you attend comments ?

**Review by hugo-cordoba — APPROVED** [review]

File: product_multi_barcode/models/product_barcode.py (hunk 6 above)
Comment (StefanRijnhart) [reply product_multi_barcode/models/product_barcode.py]: Actually, the found duplicate is used to present a detailed error message to the user so I've kept `search`.

File: product_multi_barcode/models/product_product.py (hunk 2 above)
Comment (StefanRijnhart) [reply product_multi_barcode/models/product_product.py]: I changed this, although I think write is fine in principle. The same method also uses create and unlink. But assigning a single attribute is more odoo-esque of course.

**Review by rousseldenis — APPROVED** [review]
Code review

**Comment by StefanRijnhart** [comment]
Thanks for the reviews. Fixes are now applied. I also included https://github.com/OCA/stock-logistics-barcode/pull/501

**Comment by rousseldenis** [comment]
> Thanks for the reviews. Fixes are now applied. I also included #501

@StefanRijnhart See my comment https://github.com/OCA/stock-logistics-barcode/pull/501#pullrequestreview-1370863199

**Review by aliciagaarzo — APPROVED** [review]

**Comment by StefanRijnhart** [comment]
@rousseldenis do you still have any remaining concerns, or are we ready for merge?

**Review by rousseldenis — APPROVED** [review]

**Pull request merged**